doc/api/esm.md

```diff
@@ -643,12 +643,12 @@ The resolve hook returns the resolved file URL and module format for a
 given module specifier and parent file URL:
 
 ```js
-import { URL, pathToFileURL } from 'url';
+import { pathToFileURL } from 'url';
 const baseURL = pathToFileURL(process.cwd()).href;
 
 /**
  * @param {string} specifier
- * @param {string} parentModuleURL
+ * @param {string | undefined} parentModuleURL
```

Comment:
AFAIK we never document `undefined` as valid type.

Reply:
This is the annotation that folks in the referenced issue felt was most appropriate. Please advise on how to proceed.

Reply:
I think for an example in the docs it's valuable to call out that the parent URL *may* not be passed since it might be surprising to some people. We could also put that into prose but I think I'd prefer it to be documented like is here.

Reply:
I agree with @jkrems that this should also be put into prose as what is currently stated is unclear. Here's the thing: in this particular example, there is a default defined (`baseURL`) that will be used in case `parentModuleURL` is ever undefined. This means that the value passed into the function will never actually end up being of the type `undefined` in this example. It will _always_ end up be a string no matter what. This would not be the case if that default parameter (` = baseURL`) was omitted in this example. So, we need to keep in mind that _in this item_ we are documenting _this example_ and not the behavior of _the default resolver_.

**We are documenting the _example function signature_**: 

https://github.com/nodejs/node/blob/c2807f9dc96255532b3bc744bcec59e70e36fd0c/doc/api/esm.md#L703-L705

**We are not documenting the _default resolver function signature_**: https://github.com/nodejs/node/blob/c42ab995c16a134350078d70e32bc278cd74bc1f/lib/internal/modules/esm/loader.js#L68

That being said, as of right now, I think my original suggestion it being of type `{string=}` was correct. See [this JSDoc page](https://jsdoc.app/tags-param.html#optional-parameters-and-default-values) for reference. I will be double-checking that this syntax is acceptable for both TypeScript Compiler and Closure Compiler momentarily.

Reply:
Probably worth mentioning that jsdoc param comments document the _external api_ of a function, so in this case, the second argument should be `string | undefined` as it is OK for a caller to pass `undefined` (specifically because it has a default set). A checker (like TS or closure) would then narrow that to just `string` in the body, assuming the default assignment (`baseURL`) is of type `string` and not possibly `undefined`.

Reply:
> Probably worth mentioning that jsdoc param comments document the external api of a function

@weswigham, I'm glad you clarified this, I wasn't entirely sure and I'm beginning to wonder if Closure operates the same way. I'm not doubting you, but I'm curious if this distinction is ever made in the TypeScript or Closure documentation. If you happen to know where, please share!

I think the best way forward at this point would be if I were to create a repo demonstrating valid annotations. Had I done this sooner, it might've avoided some of the disagreements found in the issue (although I did them insightful).

Reply:
This has all been addressed. Conclusion:
```
 * @param {!(URL|string|undefined)} parentModuleURL
```
Breakdown: the leading `!(...)` specifies that it will never be `null`. It must either be of type `URL`, `string`, or `undefined`. The reason it can be `URL` is that it is being fed into `URL`'s constructor on line 38 of the test fixture. The second parameter of that constructor can be either `URL` or `string`, so this variable must be one or the other (not just one). I'm not sure how ideal this is, but the compilers need it to be this way. Another compiler-related intricacy is the need for importing `{ default as Module }`, which was required by TypeScript. I also moved the assignment of the default parameter into the function. It doesn't make sense to have that as a global and it was causing the compilers to disagree with each other. I hope everyone is pleased with the result!